draft-gpew-priv-ppm.md

```diff
@@ -672,242 +679,366 @@ independently.
+state across requests. At least part of the state will usually need to be
+encrypted in order to protect user privacy. However, the details of precisely
+how the state is encrypted and the information that it carries is up to the
+helper implementation.
+
+[[OPEN ISSUE: we may end up removing helper_state. See #185]]
+
+The rest of the message is a sequence of ProcessShare values, the order of which
+matches that of the InputReportShare values in `AggregateInitReq.seq`. Each report
+that was marked as invalid is assigned the ProcessType `failed`. Otherwise, the
+ProcessShare is either marked as continued with the output `agg_msg`, or is marked
+as finished if the VDAF computation is finished.
+
+The helper's response to the leader is an HTTP 200 OK whose body is the
+AggregateInitResp and media type is "message/ppm-aggregate-init-resp".
+
+Upon receipt of a helper's AggregateInitResp message, the leader checks that the
+sequence of ProcessShare messages corresponds to the ReportShare sequence of the
+AggregateInitReq. If any message appears out of order, is missing, has an
+unrecognized nonce, or if two messages have the same nonce, then the leader MUST
+abort with error "unrecognizedMessage".
+
+### Aggregate Continuation {#agg-continue-flow}
+
+In the continuation phase, the leader drives aggregation of each share in the candidate
+report set until the underlying VDAF moves into a terminal state, which happens either
+when aggregation completes or an error occurs. The leader does this by repeating the
+following procedure.
+
+* If the helper failed, then mark the report as failed, remove it from the candidate
+  report set, and do not process it further.
+* If the leader continued but the helper finished, or if the leader finished but
+  the helper continued, then mark the report as failed and do not process it further.
+* If the leader and helper finished, then continuation is complete. At this point the
+  report has been completely processed and both aggregators have recovered an
+  output share. The leader proceeds as described in {{agg-complete}}.
+* If the leader and helper continued, then the leader computes its next state
+  transition ("prep_next") and sends the helper a ProcessShare message with the
+  payload computed as described below.
+
+The leader sends each helper a ProcessShare value for each report share that is to
+be continued. Let `leader_outbound` denote the last VDAF message the leader computed
+and let `helper_outbound` denote the last VDAF message it received from the helper.
+The payload of the the ProcessShare message is computed as
+
+~~~
+inbound = VDAF.prep_shares_to_prep(agg_param, [leader_outbound, helper_outbound])
+~~~
+
+where [leader_outbound, helper_outbound] is a vector of two elements. The next state
+transition is then computed as:
+
+~~~
+out = VDAF.prep_next(prep_state, inbound)
+~~~
+
+If either of these operations fails, then the leader marks the report as invalid.
```

Comment:
If it's invalid does it have to tell the helper?

Reply:
Our thinking here is that the leader would just skip reports that can't be processed further. The motivation for this is to keep the helper simple. It also has the advantage of avoiding communication overhead for reports that can't be aggregated. One potential downside is that the helper doesn't get to find out *why* processing failed --- VDAF evaluation may fail in different ways, for example --- whereas the leader does.

Reply:
Right now, they shouldn't actually aggregate until the leader issues an AggregateShareReq. We can change how that's done if we want to optimize things, but let's do so separately.